storey/flow.py

```diff
@@ -386,6 +388,104 @@ async def _do(self, event):
+                    for key, pending_event in self._pending_by_key.items():
+                        if self._pending_by_key[key].pending and not self._pending_by_key[key].in_flight:
+                            resp = await self._process_event(pending_event.pending[0])
+                            for event in pending_event.pending:
+                                await self._handle_completed(event, resp)
+                    if self._q.empty():
+                        break
+                    else:
+                        await self._q.put(_termination_obj)
+                        continue
+
+                event = job[0]
+                completed = await job[1]
+
+                for event in self._pending_by_key[event.key].in_flight:
+                    await self._handle_completed(event, completed)
+                self._pending_by_key[event.key].in_flight = []
+
+                # If we got more pending events for the same key process them
+                if self._pending_by_key[event.key].pending:
+                    first_event = self._pending_by_key[event.key].pending[0]
+                    self._pending_by_key[event.key].in_flight = self._pending_by_key[event.key].pending
+                    self._pending_by_key[event.key].pending = []
+
+                    task = self._process_event(first_event)
+                    await self._q.put((event, asyncio.get_running_loop().create_task(task)))
+                else:
+                    del self._pending_by_key[event.key]
+        except BaseException as ex:
+            if not self._q.empty():
+                await self._q.get()
+            raise ex
+        finally:
+            await self._cleanup()
+
+    async def _do(self, event):
+        if not self._q:
+            await self._lazy_init()
+            self._q = asyncio.queues.Queue(self._max_in_flight)
+            self._worker_awaitable = asyncio.get_running_loop().create_task(self._worker())
+
+        if self._worker_awaitable.done():
+            await self._worker_awaitable
+            raise FlowError("ConcurrentByKeyJobExecution worker has already terminated")
+
+        if event is _termination_obj:
+            await self._q.put(_termination_obj)
+            await self._worker_awaitable
+            return await self._do_downstream(_termination_obj)
+        else:
+            # Initializing the key with 2 lists. One for pending requests and one for requests that an update request has been issued for.
+            if event.key not in self._pending_by_key:
+                self._pending_by_key[event.key] = _PendingEvent()
+
+            # If there is a current update in flight for the key, add the event to the pending list. Otherwise update the key.
+            if len(self._pending_by_key[event.key].in_flight) > 0:
+                self._pending_by_key[event.key].pending.append(event)
+            else:
+                self._pending_by_key[event.key].pending.append(event)
```

Comment:
Shouldn't `self._pending_by_key[event.key].pending.append(event)` be outside the `if` if we do it either way?